src/thor/costmatrix.cc

```diff
@@ -817,6 +829,102 @@ void CostMatrix::BackwardSearch(const uint32_t index, GraphReader& graphreader)
   }
 }
 
+// Check if the edge on the forward search connects to a reached edge
+// on the reverse search trees.
+void CostMatrix::CheckReverseConnections(const uint32_t target,
+                                         const BDEdgeLabel& rev_pred,
+                                         const uint32_t n,
+                                         GraphReader& graphreader) {
+
+  // Disallow connections that are part of an uturn on an internal edge
+  if (rev_pred.internal_turn() != InternalTurn::kNoTurn) {
+    return;
+  }
+  // Disallow connections that are part of a complex restriction.
+  // TODO - validate that we do not need to "walk" the paths forward
+  // and backward to see if they match a restriction.
+  if (rev_pred.on_complex_rest()) {
+    return;
+  }
+
+  // Get the opposing edge. Get a list of source locations whose forward
+  // search has reached this edge.
+  GraphId fwd_edgeid = rev_pred.opp_edgeid();
+  auto sources = sources_->find(fwd_edgeid);
+  if (sources == sources_->end()) {
+    return;
+  }
```

Comment:
I ran the 51 locs example again multiple times each, the slowdown seems to be in the 10-15% range. 

I don't quite get it.. the code stops right here for the vast majority of reverse expanded edges, since it's only towards the end. So it's mostly "just" the lookup `sources_->find(fwd_edgeid)` (plus putting the edge/pair id into that map). I wouldn't have thought that'd correspond to ~ 10% of the whole execution time. 

\> 10% is quite a lot for a little improvement in quality. I'd also find it sad to compromise (even more) quality for performance.

WDYT @dnesbitt61 @kevinkreiser ?

Reply:
some questions.. how are you testing this? are you running valhalla_service in one shot mode or are you making http requests while keeping the server running? i always test with: https://github.com/valhalla/valhalla/blob/5b3cd9704ffd699b992b08bc4a5bf60e34d318f5/run_route_scripts/run_with_server.py

it simulates a real world condition of a long running server process with http requests hitting it in parallel. doing multiple runs of this is the most trust worthy benchmark we have (imho).

if we think its not the look up, are you sure its not the extra allocations/deallocations assocated to the new map itself. i notice that we resize to 0 and shrink to fit, essentially forcing the container to deallocate and reallocate for the next request. imho it would be better to do like we do in other places which is to allow some amount of that container to keep its reservation (ie `resize(MAX_MATRIX_CONNECTIONS_RESERVATION)` so that modestly sized matrix requests can just reuse the slots we've already allocated for previous requests.

another optimization we could do is to not use two containers. edge_ids have plenty of unused bits. you could use a single container to hold both sets of potential connections by bitwise or'ing a single bit mask over the id to differentiate between edge ids that are keys for the sources vs keys for the targets. the only intuitive benefit of this would be that regardless of search path direction it would be looking at the same bit of memory when calling `.find` on the map which might have some kind of cache hit benefit

also could we do some sanity checks? can you simply disable the call to this new function and see if all the performance is recovered? if it is then it has to either be the call to `.find` or that it gets past this point more often than we would expect and does the looping below.

long story short, can we do a little bit more investigation to inform ourselves on why? also, its totally ppossible you did all this and i just dont know it, but in case yoou didnt these are my ideas :smile:

Reply:
I used `valhalla_run_matrix` which still has the cold start problem, but with 10 runs, it shouldn't hit that hard? I'll try the run_with_server.py..

Didn't I fix those non-allocations?! I'll look into it again.

> another optimization we could do is to not use two containers. edge_ids have plenty of unused bits.

Intuitively, without looking into it, we need to use the full 46 bits for this (or whatever it was, surely more than 32). What am I missing?

I'll sanity check more, thanks for the hints!

Reply:
> Intuitively, without looking into it, we need to use the full 46 bits for this (or whatever it was, surely more than 32). What am I missing?

the key in the map is uint64_t, which is what graph_ids are. the edge ids only use up, like you said, 46 bits of that. that leaves 18 spare bits for you to do whatever you like with. to use the same container, you only need to differentiate whether you are asking the container for an edge on behalf of the forward or reverse search paths. which means you only need one bit designating forward or reverse. pick one of the directions and any time you ask for it, mask the edge id with one of the spare 18 bits flipped `lookup.find(edge_id | (1 << 47))`

Reply:
right, thanks!